### PR TITLE
Bugfix KNOWAGE-8111

### PR DESCRIPTION
**Pull request by BojanSovticEngIT**
Files changed - all except 

- knowagetalendenginclientapi/src/main/java/it/eng/spagobi/engines/talend/client/JobDeploymentDescriptor.java

 I can't find it, missing entire src folder in knowagetalendenginclientapi project.

### CODE REVIEW

File: knowage-core/src/main/java/it/eng/spagobi/engines/drivers/AbstractEngineDriver.java [hunk 1]
```diff
@@ -280,11 +283,27 @@ public JSONObject getJsonFromTemplate(byte[] contentTemplate) {
...
 
+	private static DocumentBuilderFactory getFactory() {
+		DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+		try {
+			dbf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+			dbf.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+			dbf.setFeature("http://xml.org/sax/features/external-parameterentities", false);
+			dbf.setFeature("http://xml.org/sax/features/external-generalentities", false);
+		} catch (ParserConfigurationException e) {
+			LOGGER.error("Error loading XML document: " + e.getMessage(), e);
```
Comment (davide-zerbetto) [inline knowage-core/src/main/java/it/eng/spagobi/engines/drivers/AbstractEngineDriver.java:294]: throw a wrapping SecurityException here

File: knowageutils/src/main/java/it/eng/spagobi/commons/utilities/SpagoBIUnmarshallerWrapper.java [hunk 2]
```diff
@@ -44,6 +57,28 @@ public class SpagoBIUnmarshallerWrapper<T> {
...
+		} finally {
+			try {
+				inputStream.close();
+			} catch (IOException e) {
+				LOGGER.error("Error loading XML document: " + e.getMessage(), e);
+			}
+		}
+
 		Unmarshaller unmarshaller = jc.createUnmarshaller();
 		return (T) unmarshaller.unmarshal(xmlFile);
```
Comment (davide-zerbetto) [inline knowageutils/src/main/java/it/eng/spagobi/commons/utilities/SpagoBIUnmarshallerWrapper.java:83]: Are those lines still useful?

File: knowageutils/src/main/java/it/eng/spagobi/commons/utilities/SpagoBIUnmarshallerWrapper.java [hunk 3]
```diff
@@ -44,6 +57,28 @@ public class SpagoBIUnmarshallerWrapper<T> {
...
+
+		FileInputStream inputStream = null;
+		try {
+			inputStream = new FileInputStream(xmlFile);
+			InputStreamReader reader = new InputStreamReader(inputStream, StandardCharsets.UTF_8);
+			XMLStreamReader xsr = xif.createXMLStreamReader(new StreamSource(reader));
+
+			return (T) jc.createUnmarshaller().unmarshal(xsr);
+		} catch (FileNotFoundException | XMLStreamException e) {
+			LOGGER.error("Error loading XML document: " + e.getMessage(), e);
```
Comment (davide-zerbetto) [inline knowageutils/src/main/java/it/eng/spagobi/commons/utilities/SpagoBIUnmarshallerWrapper.java:73]: Throw a wrapping runtime exception here

File: knowageutils/src/main/java/it/eng/spagobi/tools/dataset/common/datareader/XmlDataReader.java [hunk 4]
```diff
@@ -168,6 +170,23 @@ public IDataStore read(Object data) {
...
+	private static DocumentBuilderFactory getFactory() {
+		DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+		try {
+			dbf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+			dbf.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+			dbf.setFeature("http://xml.org/sax/features/external-parameterentities", false);
+			dbf.setFeature("http://xml.org/sax/features/external-generalentities", false);
+		} catch (ParserConfigurationException e) {
+			logger.error("Error loading XML document: " + e.getMessage(), e);
+			throw new SecurityException("Error loading XML document: " + e.getMessage(), e);
```
Comment (davide-zerbetto) [inline knowageutils/src/main/java/it/eng/spagobi/tools/dataset/common/datareader/XmlDataReader.java:182]: The error message is misleading: we are not loading XML document in that block

File: knowagewhatifengine/src/main/java/it/eng/spagobi/engines/whatif/calculatedmember/MDXFormulaHandler.java [hunk 5]
```diff
@@ -196,10 +205,34 @@ private static void injectTimeDimensionInBody() {
...
+		} finally {
+			try {
+				inputStream.close();
+			} catch (IOException e) {
+				LOGGER.error("Error loading XML document: " + e.getMessage(), e);
+			}
+		}
+
 		Unmarshaller unmarshaller = jc.createUnmarshaller();
 		return (T) unmarshaller.unmarshal(file);
```
Comment (davide-zerbetto) [inline knowagewhatifengine/src/main/java/it/eng/spagobi/engines/whatif/calculatedmember/MDXFormulaHandler.java:237]: Are those lines still useful?

File: knowagewhatifengine/src/main/java/it/eng/spagobi/engines/whatif/calculatedmember/MDXFormulaHandler.java [hunk 6]
```diff
@@ -196,10 +205,34 @@ private static void injectTimeDimensionInBody() {
...
+
+		FileInputStream inputStream = null;
+		try {
+			inputStream = new FileInputStream(file);
+			InputStreamReader reader = new InputStreamReader(inputStream, StandardCharsets.UTF_8);
+			XMLStreamReader xsr = xif.createXMLStreamReader(new StreamSource(reader));
+
+			return (T) jc.createUnmarshaller().unmarshal(xsr);
+		} catch (FileNotFoundException | XMLStreamException e) {
+			LOGGER.error("Error loading XML document: " + e.getMessage(), e);
```
Comment (davide-zerbetto) [inline knowagewhatifengine/src/main/java/it/eng/spagobi/engines/whatif/calculatedmember/MDXFormulaHandler.java:227]: throw a wrapping runtime exception here

File: knowage-core/src/main/java/it/eng/spagobi/engines/drivers/AbstractEngineDriver.java (hunk 1 above)
Comment (davide-zerbetto) [inline knowage-core/src/main/java/it/eng/spagobi/engines/drivers/AbstractEngineDriver.java:294]: The error message is misleading: we are not loading XML document in that block

File: knowage-core/src/main/java/it/eng/spagobi/engines/drivers/chart/ChartDriver.java [hunk 7]
```diff
@@ -224,6 +226,23 @@ private Map applyService(Map parameters, BIObject biObject) {
...
 
+	private static DocumentBuilderFactory getFactory() {
+		DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+		try {
+			dbf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+			dbf.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+			dbf.setFeature("http://xml.org/sax/features/external-parameterentities", false);
+			dbf.setFeature("http://xml.org/sax/features/external-generalentities", false);
+		} catch (ParserConfigurationException e) {
+			logger.error("Error loading XML document: " + e.getMessage(), e);
```
Comment (davide-zerbetto) [inline knowage-core/src/main/java/it/eng/spagobi/engines/drivers/chart/ChartDriver.java:237]: The error message is misleading: we are not loading XML document in that block

File: knowagetalendengine/src/main/java/it/eng/spagobi/engines/talend/runtime/JobDeploymentDescriptor.java [hunk 8]
```diff
@@ -1,122 +1,134 @@
...
-	    }
+
+		try {
+			reader.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+			reader.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+			reader.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
+			reader.setFeature("http://xml.org/sax/features/external-general-entities", false);
+		} catch (SAXException | SecurityException e) {
+			LOGGER.error("Error loading XML document: " + e.getMessage(), e);
+			throw new SecurityException("Error loading XML document: " + e.getMessage(), e);
```
Comment (davide-zerbetto) [inline knowagetalendengine/src/main/java/it/eng/spagobi/engines/talend/runtime/JobDeploymentDescriptor.java:79]: The error message is misleading: we are not loading XML document in that block

**Review by davide-zerbetto — CHANGES_REQUESTED** [review]

File: knowagewhatifengine/src/main/java/it/eng/spagobi/engines/whatif/calculatedmember/MDXFormulaHandler.java (hunk 5 above)
Comment (BojanSovticEngIT) [reply knowagewhatifengine/src/main/java/it/eng/spagobi/engines/whatif/calculatedmember/MDXFormulaHandler.java:237]: I think so, I think it is used for creating Java object from parsed XML.

https://www.tabnine.com/code/java/methods/javax.xml.bind.Unmarshaller/unmarshal

I'm working on other changes.

File: knowagetalendengine/src/main/java/it/eng/spagobi/engines/talend/runtime/JobDeploymentDescriptor.java (hunk 8 above)
Comment (BojanSovticEngIT) [reply knowagetalendengine/src/main/java/it/eng/spagobi/engines/talend/runtime/JobDeploymentDescriptor.java]: Fixed

File: knowage-core/src/main/java/it/eng/spagobi/engines/drivers/chart/ChartDriver.java (hunk 7 above)
Comment (BojanSovticEngIT) [reply knowage-core/src/main/java/it/eng/spagobi/engines/drivers/chart/ChartDriver.java]: Fixed

File: knowage-core/src/main/java/it/eng/spagobi/engines/drivers/AbstractEngineDriver.java (hunk 1 above)
Comment (BojanSovticEngIT) [reply knowage-core/src/main/java/it/eng/spagobi/engines/drivers/AbstractEngineDriver.java:294]: Fix

File: knowageutils/src/main/java/it/eng/spagobi/tools/dataset/common/datareader/XmlDataReader.java (hunk 4 above)
Comment (BojanSovticEngIT) [reply knowageutils/src/main/java/it/eng/spagobi/tools/dataset/common/datareader/XmlDataReader.java]: Fixed

File: knowagewhatifengine/src/main/java/it/eng/spagobi/engines/whatif/calculatedmember/MDXFormulaHandler.java (hunk 6 above)
Comment (BojanSovticEngIT) [reply knowagewhatifengine/src/main/java/it/eng/spagobi/engines/whatif/calculatedmember/MDXFormulaHandler.java:227]: Possibly fixed.

File: knowagewhatifengine/src/main/java/it/eng/spagobi/engines/whatif/calculatedmember/MDXFormulaHandler.java (hunk 5 above)
Comment (BojanSovticEngIT) [reply knowagewhatifengine/src/main/java/it/eng/spagobi/engines/whatif/calculatedmember/MDXFormulaHandler.java:237]: Possibly fixed (with issue below).

File: knowageutils/src/main/java/it/eng/spagobi/commons/utilities/SpagoBIUnmarshallerWrapper.java (hunk 3 above)
Comment (BojanSovticEngIT) [reply knowageutils/src/main/java/it/eng/spagobi/commons/utilities/SpagoBIUnmarshallerWrapper.java:72]: Possibly fixed.

File: knowageutils/src/main/java/it/eng/spagobi/commons/utilities/SpagoBIUnmarshallerWrapper.java (hunk 2 above)
Comment (BojanSovticEngIT) [reply knowageutils/src/main/java/it/eng/spagobi/commons/utilities/SpagoBIUnmarshallerWrapper.java]: Possibly fixed (with issue below).

File: knowage-core/src/main/java/it/eng/spagobi/engines/drivers/AbstractEngineDriver.java (hunk 1 above)
Comment (BojanSovticEngIT) [reply knowage-core/src/main/java/it/eng/spagobi/engines/drivers/AbstractEngineDriver.java]: Possibly fixed.

**Pull request merged**